Updates for effector calling.

geomacy opened the pull request:
this includes the following changes:

An "avoidSideEffects" capability in ValueResolver, to allow callers to
specify that any tasks which are marked as having side effects (a new
marker Interface) should be ignored.  The use case for this is illustrated
by its use here in  AbstractConfigurationSupportInternal in the
getNonBlocking method invoked by ConfigConstraints.validateAll.  This is
called during validation of the YAML, and evaluates all config key values,
which without this change would result in unwanted repeat executions of
effectors (and at a point when their target entities have not yet been started).
Similarly it is called during calculation of a hash for the extra salt for
VanillaSoftwareProcessSshDriver, again a situation where we do not want
effectors being called.

An additional effector() method on BrooklynDslCommon, to add support for
varArgs passing of the effector arguments, and a matching new method on
BrooklynDslDeferredSupplier.

The task created by ExecuteEffector is cached, in order to avoid unwanted
repeated executor invocations,  as illustrated in the test
testEffectorCalledOncePerConfigKey in EffectorsYamlTest.

test-app-with-effectors.yaml is removed as it's easier to see what's going
on with the YAML inline within the tests.

New tests are added in EffectorsYamlTest.

A new 'sequenceEffector' is added, just added directly  to TestEntity, in case
it happens to be useful for anyone else.

geomacy committed:
Updates for effector calling.

this includes the following changes:

An "avoidSideEffects" capability in ValueResolver, to allow callers to
specify that any tasks which are marked as having side effects (a new
marker Interface) should be ignored.  The use case for this is illustrated
by its use here in  AbstractConfigurationSupportInternal in the
getNonBlocking method invoked by ConfigConstraints.validateAll.  This is
called during validation of the YAML, and evaluates all config key values,
which without this change would result in unwanted repeat executions of
effectors (and at a point when their target entities have not yet been started).
Similarly it is called during calculation of a hash for the extra salt for
VanillaSoftwareProcessSshDriver, again a situation where we do not want
effectors being called.

An additional effector() method on BrooklynDslCommon, to add support for
varArgs passing of the effector arguments, and a matching new method on
BrooklynDslDeferredSupplier.

The task created by ExecuteEffector is cached, in order to avoid unwanted
repeated executor invocations,  as illustrated in the test
testEffectorCalledOncePerConfigKey in EffectorsYamlTest.

test-app-with-effectors.yaml is removed as it's easier to see what's going
on with the YAML inline within the tests.

New tests are added in EffectorsYamlTest.

A new 'sequenceEffector' is added, just added directly  to TestEntity, in case
it happens to be useful for anyone else.

aledsage commented on camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/DslComponent.java:
```diff
@@ -255,8 +271,14 @@ public ExecuteEffector(DslComponent component, String effectorName, Map<String,
             }
-            return (Task<Object>) Entities.invokeEffector(targetEntity, targetEntity, targetEffector.get(), args);
+            if (null == cachedTask) {
+                cachedTask = null == argList
```
Preference for brackets around `(null -- argList)`.

aledsage replied:
We also need to think about the resolved value surviving rebind. We probably want to store the result somehow, as well as the task (because persistence won't cope well with tasks).

aledsage commented on camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/EffectorsYamlTest.java:
```diff
@@ -18,45 +18,168 @@
+            "  id: entity1",
+            "  brooklyn.config:",
+            "    test.confName: ",
+            "      $brooklyn:entity(\"entity1\").effector(\"identityEffector\", \"hello\")"
```
Worth noting that the multi-line version that took args doesn't work (due to our DSL parsing we believe). And linking to a brooklyn jira issue, once you've created that.

geomacy replied:
I will add in a test that doesn't work and mark it as disabled, and raise the JIRA.

aledsage commented on camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/EffectorsYamlTest.java:
```diff
@@ -18,45 +18,168 @@
+    }
+
+    @Test(groups = "Integration")
+    public void testSshCommandSensorWithEffectorInEnv() throws Exception {
```
Personal preference for keeping the integration tests like this in a separate class. We can also use a `RecordingSshTool` so that it avoids any ssh commands, and thus keep it as a unit test.

aledsage commented on core/src/main/java/org/apache/brooklyn/util/core/task/ValueResolver.java:
```diff
@@ -203,6 +205,11 @@ public ResolverBuilderPretype(Object v) {
         return this;
     }
+
+    public ValueResolver<T> avoidSideEffects() {
```
We should write a unit test for `ValueResolver`, as well as testing it in the yaml files.

geomacy committed:
Adding integration test for entity interaction.

Also adds a formatString test that was used to get the right
structure for the entity test, may as well include it now
that it is written.
